cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -448,24 +543,55 @@ public boolean apply(RunRecordMeta input) {
     return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_STARTED, limit);
   }
 
+  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(Set<ProgramRunId> programRunIds, int limit) {
+    return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_COMPLETED, limit);
+  }
+
   private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(Set<ProgramRunId> programRunIds, String recordType,
                                                               int limit) {
     Set<MDSKey> keySet = new HashSet<>();
     for (ProgramRunId programRunId : programRunIds) {
       keySet.add(getProgramKeyBuilder(recordType, programRunId).build());
     }
     Map<MDSKey, RunRecordMeta> returnMap = listKV(keySet, RunRecordMeta.class, limit);
-    return getProgramRunIdMap(returnMap);
-  }
-
-  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(Set<ProgramRunId> programRunIds, int limit) {
-    Set<MDSKey> keySet = new HashSet<>();
+    Map<ProgramRunId, RunRecordMeta> programRunIdRunRecordMetaMap = getProgramRunIdMap(returnMap);
+    Set<ProgramRunId> candidateSet = Sets.newHashSet(programRunIdRunRecordMetaMap.keySet());
+    for (ProgramRunId programRunId : candidateSet) {
+      // If the program run id, doesn't match with the requested set, remove it.
+      if (!programRunIds.contains(programRunId)) {
+        programRunIdRunRecordMetaMap.remove(programRunId);
+      }
+    }
 
-    for (ProgramRunId programRunId : programRunIds) {
-      keySet.add(getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programRunId).build());
+    int remaining = limit - programRunIdRunRecordMetaMap.size();
+    if (remaining > 0) {
+      keySet.clear();
+      for (ProgramRunId programRunId : programRunIds) {
+        if (programRunId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+          keySet.add(getVersionLessProgramKeyBuilder(recordType, programRunId).build());
+        }
+      }
+      returnMap = listKV(keySet, RunRecordMeta.class, remaining);
+      Map<ProgramRunId, RunRecordMeta> versionLessMap = getProgramRunIdMap(returnMap);
+      for (Map.Entry<ProgramRunId, RunRecordMeta> oldFormatEntry : versionLessMap.entrySet()) {
+        // Skip the record if it is present in the new format
+        if (programRunIdRunRecordMetaMap.keySet().contains(oldFormatEntry.getKey())) {
+          continue;
+        }
+
+        // Skip the record if it is not present in the requested set
+        if (!programRunIds.contains(oldFormatEntry.getKey())) {
```

Comment:
Yes because the way the row keys are organized we have programId:startTs:runId. So when we query for some ProgramRunIds, it is hard to perform that scan and thus we have to scan based on programId and then pick only the runIds that match.